### PR TITLE
0.5.3

### PR DESCRIPTION
**Pull request by 3dg1luk43**
0.5.3

### CODE REVIEW

File: custom_components/ha_washdata/playground.py [hunk 1]
```diff
@@ -860,6 +913,106 @@ def run_tail(self) -> None:
...
+            synthetic_from_s = self.cursor["t"]
+            stop_thresh = float(getattr(self.config, "stop_threshold_w", 2.0))
+            idle_above = idle_w >= stop_thresh
+
+            seed = sum(ord(c) for c in (self.cycle.get("id") or "pg")) & 0xFFFF
+
+            self._emit(
+                "stress_tail_start",
+                f"idle ~{idle_w:.1f}W fluct={fluct_w:.2f}W above_stop_thresh={idle_above}",
+            )
```
[the remark this answers is not in the document]
Comment (3dg1luk43) [reply custom_components/ha_washdata/playground.py:991]: Addressed in the follow-up fix commit.

The three user-facing stress alert details (`stress_terminated`, `stress_above_threshold`, `stress_hit_cap`) now go through the panel's `detail_key` / `detail_params` mechanism (the same path already used for cycle-artifact details at `ha-washdata-panel.js`): each alert carries `detail_key` + `detail_params` with the English text kept as the fallback, the keys are added to `translations/panel/en.json` (`msg.pg_stress_*_detail`) and translated into all 34 languages, and the alert renderer now resolves `detail_key` via `_t(...)`. The hardcoded "8 h" in `stress_hit_cap` was also replaced with the actual elapsed duration.

The `stress_tail_start` **event** detail (`idle ~{}W fluct={}W above_stop_thresh={}`) is intentionally left as a raw string: it is diagnostic telemetry (a variable dump) shown dimmed in the event hover tooltip, and every playground event detail emitted via `_emit(...)` is a raw diagnostic string of this kind (`state`, `detected`, `finished`, `match_commit`, `match_ambiguous`, `unmatched`, ...). It is not localized UI chrome, and localizing a single event's variable dump while the rest stay raw would be inconsistent and add churn without user value.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

File: custom_components/ha_washdata/playground.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (3dg1luk43) [reply custom_components/ha_washdata/playground.py:991]: Thanks for narrowing the finding on the event-detail telemetry.

The three keys are present in the current PR head (5388595):
`msg.pg_stress_terminated_detail`, `msg.pg_stress_above_threshold_detail`, and `msg.pg_stress_hit_cap_detail` are in `translations/panel/en.json` and in all 34 other `translations/panel/*.json` files (35 JSON files total, verified with `git grep -l ... HEAD`). The prior review snapshot predated the commit, hence the miss. Re-triggering so this can be re-verified against the current head.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Thanks for the follow-up pass. Addressed in commit 1108c24:

**Fixed**
- **playground override domain** (playground.py): non-finite (`nan`/`inf`) overrides are now rejected (auto-derived floor kept) and finite values clamped to `[0, PLAYGROUND_STRESS_MAX_IDLE_W]`, so nothing absurd can reach the synthetic samples. Regression tests added (non-finite + huge).
- **cap label vs dynamic detail** (`lbl.pg_alert_stress_cap`): the label is now the generic "Hit safety cap" (en + all 34 languages), so it no longer contradicts the dynamic `h`/`m` in the detail.
- **`include_pending`/`page_size` keyword-only** on `list_brands` and `search_devices` (resolves the FBT001/FBT002 flags on the changed signatures).
- **forced-stop series point** (playground.py): a `_sample()` is now recorded at the force-stop timestamp so the plotted tail reaches the reported elapsed time.
- **catalog completeness** (store_client.py): both `list_brands` and `search_devices` now cache the full catalog (ceiling raised to 500, far above the real catalog) and filter the prefix in memory over the complete set, so entries past the old 60-row limit stay searchable for the whole TTL. A query still reads only the docs that exist, so no added reads for today's catalog.
- **cache invalidation** (store_client.py): `confirm_device` now invalidates the catalog cache when it promotes a device `pending -> approved`, and every catalog read now captures a generation counter and only caches its result if no invalidation occurred while the query was in flight (closes the fill-after-invalidate race).
- **replace-mode wipe** (profile_store.py): the replace wipe now targets the list each ticked cycle category actually writes to, so `real_cycles` routed to the default reference destination wipes `reference_cycles` too (not only when the `reference_cycles` category itself is ticked). Regression test added.

**Intentionally not changed (rationale)**
- **`_derive_idle_level` gap-awareness** (playground.py): this returns a p7 (7th-percentile) standby *floor*, an order statistic that selects the low/between-burst readings and is inherently robust to how samples cluster in time. It is a diagnostic seed for a what-if replay, not a persisted energy/time integration, so the sample-count percentile is appropriate here (the guideline targets `Σ P·dt`). A naive contiguous-tail split would also misread change-only sensors, where a gap signals a *stable* (constant) power level i.e. standby, not an outage, and would discard valid standby readings. `fluct_w` is only cosmetic jitter for the synthetic tail.
- **re-simulate on Apply** (ha-washdata-panel.js): commit 1ffd653 deliberately made the detail sim run only on an explicit Run, removing every `_pgRerunDetail()` auto-trigger including the one in `_pgApplySweepValue`, to avoid re-running the heavy replay on each parameter change. After an Apply the threshold lines/inputs reflect the new value and the user presses Run to see the new outcome; re-adding an auto-rerun here would revert that intentional design. Flagging for the maintainer to decide rather than overriding it.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Round-3 items addressed in commit 6e2f1e1:

**Fixed**
- **filtered replacement guard** (profile_store.py): the per-item id filters are now applied *before* the replace-mode guard, and the guard rejects a replace whose SELECTED (post-filter) cycle set is empty, so a stale/malformed id selection aborts instead of wiping a destination empty. Steps 2/3 reuse the selected sets. Regression test added.
- **manual_override provenance** (playground.py): now `True` only when a finite override was actually applied; a non-finite override that falls back to the derived floor reports `False`. Test added.
- **panel idle input** (ha-washdata-panel.js): uses `Number.isFinite` and clamps to `>= 0`, so `Infinity` (`parseFloat("1e999")`) and negatives are rejected at the input.
- **settings-revert command type** (ha-washdata-panel.js): fixed `ws_set_options` -> `set_options` (the registered command), so the revert no longer hits `unknown_command`.
- **cache helper** (store_client.py): extracted `_cached_catalog_query`, so the cache-get / generation-capture / query / conditional-store flow (including the in-flight generation guard) lives in one place, shared by `list_brands` and `search_devices`.

**Intentionally not changed (rationale)**
- **NumPy for `math.sin`/`sorted`/manual variance** (playground.py): the repo rule is "NumPy only - no SciPy, scikit-learn, or other ML libraries", i.e. it bans extra numeric/ML *libraries*, not the Python stdlib `math` module. `math` is imported and used across 6 integration modules (cycle_detector, progress, phase_match, manager, profile_store, playground). Scalar `math.sin`/`math.isfinite` are faster than their NumPy equivalents, and `sorted()` over a handful of trailing readings needs no ndarray. No SciPy/sklearn is introduced, so the runtime-dependency rule holds.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Round-4 items addressed in commit 68c1842:

**Fixed**
- **residual usability gap in the replace guard** (profile_store.py): extracted `_usable_reference_pairs` (now the single source of the "is this trace importable" rule, shared by `_add_reference_cycle_nosave` and the guard). The replace guard now requires at least one *usable* trace for reference-bound cycles (>= 2 finite points, positive span), so a selection whose cycles all carry degenerate traces aborts instead of wiping the destination and then silently adding nothing. Regression test added.
- **coalesce concurrent cache misses** (store_client.py): `_cached_catalog_query` now keeps a per-key in-flight `Future`; concurrent panel misses for the same key await the single in-flight query instead of each issuing a Firestore read. This complements the caching (fewer reads on the free-tier budget).

**Intentionally not changed (rationale)**
- **"keep catalog processing local" / "No external API calls"**: `store_client.py` *is* the community-store network client - an explicitly opt-in online feature for browsing/sharing community device profiles. The "all processing must be local" rule governs cycle detection and ML inference (which never phone home); it does not apply to the online catalog, whose entire purpose is talking to the community backend. The Firestore reads here predate this PR; this PR only *added caching* on top, which reduces those calls.
- **remove the 500 ceiling / paginate until exhausted**: 500 is a deliberate practical full-catalog bound. The community catalog holds on the order of dozens of brands and dozens of models per brand+type, so a single limited query already returns the *complete* catalog in one read, and a query only reads the docs that exist. Paginating until exhausted would directly defeat this PRs purpose - it exists to *cut* free-tier Firestore reads, and an unbounded page loop would issue many reads per cache-miss. If the catalog ever approaches the ceiling, the correct fix is a server-side prefix filter, not client-side fetch-everything; flagging for the maintainer rather than reversing the read-reduction goal here.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Round-5 items addressed in commit accf45c:

**Fixed**
- **malformed trace row** (profile_store.py `_usable_reference_pairs`): now rejects non-list `points` and non-list/tuple rows before `len()`, so a scalar row (e.g. `[[0,10], 1]`) makes the trace unusable instead of raising `TypeError`.
- **non-corrupting replace under malformed input** (profile_store.py): addressed the root cause of the wipe-then-fail concern directly - (a) `_add_reference_cycle_nosave` now ignores a malformed `sampling_interval` (try/except) rather than raising after the wipe; (b) the real-history path skips records missing `start_time`/`duration` (which `_add_cycle_data` requires); and (c) each per-record add in Steps 2/3 is wrapped so a single malformed record is skipped-and-logged, never aborting the loop. Combined with the pre-wipe usability guard (which ensures at least one importable record), a replace can no longer wipe a destination and then leave it partially filled or empty. I chose best-effort-per-record over a full staged copy: it is non-corrupting (each record fully applies or is skipped, so `self._data` is always left consistent), and for a replace it is preferable to import the valid records and drop the malformed ones rather than abort the whole operation.
- **single-flight hardening** (store_client.py): the in-flight map now stores `(generation, task)`. A caller only joins an in-flight query of the *current* generation, so a query started before `_invalidate_catalog_cache()` is not joined by a post-invalidation caller (which would receive a pre-write snapshot) - that caller starts a fresh query. Waiters `await asyncio.shield(task)`, so one waiter cancelling neither cancels the shared query nor the others coalesced onto it; the shared body (`_fetch_and_cache`) clears its slot only if it is still the registered task. Coalescing test added.

**Intentionally not changed (rationale)**
- **`np.isfinite` instead of `math.isfinite`**: same as the earlier note - the repo rule is "NumPy only - no SciPy, scikit-learn, or other ML libraries", i.e. it bans extra numeric/ML *libraries*, not the Python stdlib `math` module. `math` is imported and used across 6 integration modules (cycle_detector, progress, phase_match, manager, profile_store, playground). `math.isfinite` on a scalar is faster than `np.isfinite` and consistent with the rest of the codebase; no SciPy/sklearn is introduced.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Addressed the follow-up finding in commit f1f354e:

- **canonicalize reference provenance** (profile_store.py): added `_bare_store_id`, which strips an existing `store:` prefix before re-prefixing. A reference cycle exported after a prior import carries `meta.source="store:<id>"`; the import was re-prefixing it to `store:store:<id>` (and accumulating another `store:` per round-trip), so the dedup key and the stamped source drifted apart. Now both the dedup key and the stored `meta.source` are the canonical single-prefix form, so re-imports and round-trips dedup correctly instead of double-weighting the envelope. Regression test added (`test_reference_reimport_canonical_source_idempotent`).

The two remaining open threads are the earlier rationale items (reiterated, not changed):
- **`math.isfinite` vs `np.isfinite`**: the "NumPy only" rule bans extra numeric/ML *libraries* (SciPy/sklearn), not the Python stdlib `math` module, which is used across 6 integration modules; a scalar `math.isfinite` is faster and consistent with the codebase, and introduces no new dependency.
- **"keep catalog processing local"**: `store_client.py` is the opt-in *online* community-catalog client; the local-only rule governs cycle detection/ML (which never phone home), not the catalog whose purpose is to talk to the community backend. This PR only added caching, which reduces those reads.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Both nits addressed in commit a3d86ce:
- **Ruff SIM105**: the malformed-`sampling_interval` guard now uses `contextlib.suppress(TypeError, ValueError)`.
- **skipped-record visibility**: the import summary now includes `skipped_cycles`, counting any record dropped by the per-record safety net, so skips are no longer only in debug logs.

The two remaining open threads are the previously-explained rationale items (unchanged): stdlib `math.isfinite` (the "NumPy only" rule bans scipy/sklearn, not the `math` module), and "keep processing local" applied to `store_client.py` (the opt-in online community-catalog client). Deferring those to your judgement.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Addressed in commit 96d8d16: `_bare_store_id` now strips every leading `store:` prefix, and `existing_ref_sources` is built via the canonical dedup key, so a legacy double-prefixed persisted value (`store:store:<id>`) and a fresh canonical import (`store:<id>`) collapse to the same key and dedup correctly. Regression test added. The two remaining open threads are the previously-explained rationale items (stdlib `math.isfinite`; `store_client.py` being the opt-in online catalog client), unchanged.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Addressed in commit d704ff5: added `_any_real_importable` (the real-history analog of the reference `_any_usable` guard) so a replace into real_history aborts unless at least one selected record has both `start_time` and `duration` (the fields `_add_cycle_data` requires and that Step 3 skips on) - it can no longer wipe `past_cycles` and refill nothing. Regression test added. Remaining open threads are the two prior rationale items (stdlib `math.isfinite`; `store_client.py` = opt-in online catalog), unchanged.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
Addressed in commit ce59208: building `existing_ref_sources` now guards with `isinstance(meta, dict)`, so a persisted reference cycle with a non-dict `meta` (legacy/raw import) no longer raises `AttributeError` and blocks the import. Regression test added. Remaining open threads are the two prior rationale items (stdlib `math.isfinite`; `store_client.py` = opt-in online catalog client), unchanged.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review

File: custom_components/ha_washdata/manager.py [hunk 2]
```diff
@@ -1858,40 +1875,42 @@ async def async_reload_config(self, config_entry: ConfigEntry) -> None:
...
+                if state and state.state not in (STATE_UNKNOWN, STATE_UNAVAILABLE):
+                    try:
+                        power = float(state.state)
+                        self.detector.process_reading(power, dt_util.now())
+                    except ValueError:
+                        self._logger.debug(
+                            "Initial power value for %s after config reload is not numeric: %r",
+                            self.power_sensor_entity_id,
+                            state.state,
+                        )
```
[the remark this answers is not in the document]
Comment (3dg1luk43) [reply custom_components/ha_washdata/manager.py:1912]: @coderabbitai The ordering is intentional. The initial read of the new sensor only runs when  is **not** in  (i.e. the appliance is idle). In that state the thresholds are irrelevant to the classification of the single kick-start reading: any non-zero watt value just primes the EMA buffer; the detector is not deciding start/stop here. If the user changes both the sensor and a threshold in the same save, the first sample goes through the old thresholds but the detector is idle so there is no wrong start/stop decision. The very next real state change from the new sensor uses the updated config. Restructuring async_reload_config to capture the initial reading after the config sync would require either storing the power value across the function boundary or a second async executor round-trip -- a significant complexity increase for a narrow, self-correcting edge case with no observable user impact. Leaving as-is.

File: custom_components/ha_washdata/store_client.py [hunk 3]
```diff
@@ -241,13 +296,33 @@ async def _run_query(self, sq: dict[str, Any], parent: str = "") -> list[dict[st
...
+        cached = self._cache_get(key)
+        if cached is not None:
+            return cached
+        gen = self._cache_gen
+        fetched = await self._run_query(build_sq())
+        if fetched is None:
+            return []
+        if gen == self._cache_gen:
+            self._cache_put(key, fetched, self._CATALOG_CACHE_TTL_S)
+        return fetched
```
[the remark this answers is not in the document]
Comment (3dg1luk43) [reply custom_components/ha_washdata/store_client.py]: @coderabbitai The "no external API calls" guideline governs the integration core (detection, matching, ML, entity updates) -- code that must work fully offline. `store_client.py` is the opt-in online community catalog and is explicitly gated (`community_catalog_enabled` option, off by default). The entire purpose of this module is to query the Firestore catalog; a local-only implementation of it would have no function. This is the same rationale posted in the previous review round. Closing as false positive.

**Comment by 3dg1luk43** [comment]
@coderabbitai review

**Comment by 3dg1luk43** [comment]
@coderabbitai review
